add public address handling to PNP

fviale opened the pull request:
- Added proactive.net.public_address setting. It allows to specify a hostname or ip address which can be used to contact the Remote Objects.
 - This configuration is implemented in the PNP protocol only. In case this setting is configured, the PNP protocol creates urls like pnp://public_address@private_address:port
 - To contact a Remote Object which contains a public address, the pnp protocol tries first to open a socket to the private address, and if the connection fails, it will create a socket to the public address.
 - URIBuilder methods now have userInfo and query parameters
 - Better escaping of urls in jmx FactoryName
 - added a functional test TestPNPPublicAddress

fviale committed:
add public address handling to PNP

 - Added proactive.net.public_address setting. It allows to specify a hostname or ip address which can be used to contact the Remote Objects.
 - This configuration is implemented in the PNP protocol only. In case this setting is configured, the PNP protocol creates urls like pnp://public_address@private_address:port
 - To contact a Remote Object which contains a public address, the pnp protocol tries first to open a socket to the private address, and if the connection fails, it will create a socket to the public address.
 - URIBuilder methods now have userInfo and query parameters
 - Better escaping of urls in jmx FactoryName
 - added a functional test TestPNPPublicAddress

vinseon commented on programming-extensions/programming-extension-pnp/src/main/java/org/objectweb/proactive/extensions/pnp/PNPRemoteObjectFactoryBackend.java:
```diff
@@ -230,7 +231,7 @@ public InternalRemoteRemoteObject createRemoteObject(RemoteObject<?> remoteObjec
 
             URI uri = new URI(this.getProtocolId(),
-                              null,
+                              ProActiveInet.getPublicAddress(),
```
I think we should explicitly indicate here that the public IP address is intentionally given through the "UserInfo" field and explain why (namespace issue using the 'query' field ?).

vinseon reviewed (changes requested):
Nice improvements overall

fviale commented:
Retest this please

paraita reviewed (approved):
Works but I had to run the spotlessApply task on my machine for the tests to pass

fviale committed:
add public address handling to PNP - reviews
- added a comment which describes the use of userinfo in the pnp uri
- use override annotation
